extras/gizmo/axis-shapes.js

```diff
@@ -0,0 +1,650 @@
+    const options = {
+        positions: [],
+        normals: [],
+        uvs: [],
+        indices: [],
+        colors: []
+    };
+
+    mesh.getPositions(options.positions);
+    mesh.getNormals(options.normals);
+    mesh.getIndices(options.indices);
+    mesh.getUvs(0, options.uvs);
+
+    const wtm = entity.getWorldTransform().clone().invert();
+    tmpV1.copy(LIGHT_DIR);
+    wtm.transformVector(tmpV1, tmpV1);
+    tmpV1.normalize();
+    const numVertices = mesh.vertexBuffer.numVertices;
+    calculateShadowColors(tmpV1, numVertices, options.normals, options.colors);
+
+    return createMesh(device, options.positions, options);
+}
+
+function calculateShadowColors(lightDir, numVertices, normals, colors = []) {
+    for (let i = 0; i < numVertices; i++) {
+        const x = normals[i * 3];
+        const y = normals[i * 3 + 1];
+        const z = normals[i * 3 + 2];
+        tmpV2.set(x, y, z);
+
+        const dot = lightDir.dot(tmpV2);
+        const shadow = dot * SHADOW_DAMP_SCALE + SHADOW_DAMP_OFFSET;
+        colors.push(shadow * 255, shadow * 255, shadow * 255, 1);
+    }
+
+    return colors;
+}
+
+class AxisShape {
+    _position;
+
+    _rotation;
+
+    _scale;
+
+    _layers = [];
+
+    _defaultColor;
+
+    _hoverColor;
+
+    device;
+
+    axis;
+
+    entity;
+
+    meshTriDataList = [];
+
+    meshInstances = [];
+
+    constructor(device, options) {
+        this.device = device;
+        this.axis = options.axis ?? 'x';
+        this._position = options.position ?? new Vec3();
+        this._rotation = options.rotation ?? new Vec3();
+        this._scale = options.scale ?? new Vec3(1, 1, 1);
+
+        this._layers = options.layers ?? this._layers;
+
+        this._defaultColor = options.defaultColor ?? Color.BLACK;
+        this._hoverColor = options.hoverColor ?? Color.WHITE;
+    }
+
+    hover(state) {
+        const material = state ? this._hoverColor : this._defaultColor;
+        for (let i = 0; i < this.meshInstances.length; i++) {
+            this.meshInstances[i].material = material;
+        }
+    }
+}
+
+class AxisArrow extends AxisShape {
+    _gap = 0;
+
+    _lineThickness = 0.04;
+
+    _lineLength = 0.5;
+
+    _arrowThickness = 0.15;
+
+    _arrowLength = 0.2;
+
+    _intersectTolerance = 0.1;
+
+    constructor(device, options = {}) {
+        super(device, options);
+
+        this.meshTriDataList = [
+            new MeshTriData(createCone(this.device)),
+            new MeshTriData(createCylinder(this.device))
+        ];
+
+        this._createArrow();
+    }
+
+    set gap(value) {
+        this._gap = value ?? 0;
+        this._updateHead();
+        this._updateLine();
+    }
+
+    get gap() {
+        return this._gap;
+    }
+
+    set lineThickness(value) {
+        this._lineThickness = value ?? 1;
+        this._updateHead();
+        this._updateLine();
+    }
+
+    get lineThickness() {
+        return this._lineThickness;
+    }
+
+    set lineLength(value) {
```

Comment:
default value to 1, instead of doing `?? 1` on the next line for all these setters.

Reply:
I don't think setters support default parameters.

Reply:
ah ok